File: Editorial/98_Validate_Binary_Search_Tree/CSharp/2_Iterative_Traversal_with_Valid_Range.cs
public class Solution {
    private Stack<TreeNode> stack = new Stack<TreeNode>();
    private Stack<int?> lowerLimits = new Stack<int?>();
    private Stack<int?> upperLimits = new Stack<int?>();

    public void Update(TreeNode root, int? low, int? high) {
        stack.Push(root);
        lowerLimits.Push(low);
        upperLimits.Push(high);
    }

    public bool IsValidBST(TreeNode root) {
        int? low = null, high = null;
        Update(root, low, high);
        while (stack.Count > 0) {
            root = stack.Pop();
            low = lowerLimits.Pop();
            high = upperLimits.Pop();
            if (root == null)
                continue;
            int val = root.val;
            if (low != null && val <= low) {
                return false;
            }

            if (high != null && val >= high) {
                return false;
            }

            Update(root.right, val, high);
            Update(root.left, low, val);
        }

        return true;
    }
}

File: Editorial/98_Validate_Binary_Search_Tree/TypeScript/2_Iterative_Traversal_with_Valid_Range.ts
function isValidBST(root: TreeNode | null): boolean {
    const stack: (TreeNode | null)[] = [],
        lowerLimits: (number | null)[] = [],
        upperLimits: (number | null)[] = [];
    function update(
        node: TreeNode | null,
        low: number | null,
        high: number | null,
    ) {
        stack.push(node);
        lowerLimits.push(low);
        upperLimits.push(high);
    }
    update(root, null, null);
    while (stack.length > 0) {
        let node = stack.pop()!;
        let low = lowerLimits.pop();
        let high = upperLimits.pop();
        if (node === null) continue;
        let val = node.val;
        if (low !== null && val <= low) {
            return false;
        }
        if (high !== null && val >= high) {
            return false;
        }
        update(node.right, val, high);
        update(node.left, low, val);
    }
    return true;
}
